source/Calamari.Common/Features/Scripting/Bash/Bootstrap.sh

```diff
@@ -263,4 +263,61 @@ function log_environment_information
 	echo "##octopus[stdout-default]"
 }
 
-log_environment_information
\ No newline at end of file
+log_environment_information
+
+function decrypt_and_parse_variables {
+    local encrypted="$1"
+    local iv="$2"
+
+    local decrypted
+    decrypted=$(decrypt_variable "$encrypted" "$iv")
+    declare -gA octopus_parameters=()
+
+    local -a key_byte_lengths=()
+    local -a value_byte_lengths=()
+    local -a hex_parts=()
+
+    while IFS='$' read -r hex_key hex_value; do
+        hex_value="${hex_value//$'\n'/}"
+        key_byte_lengths+=( $(( ${#hex_key} / 2 )) )
+        value_byte_lengths+=( $(( ${#hex_value} / 2 )) )
+        hex_parts+=( "${hex_key}${hex_value}" )
+    done <<< "$decrypted"
+
+    local concatenated_hex
+    concatenated_hex=$(printf "%s" "${hex_parts[@]}")
+
+    exec 3< <(echo -n "$concatenated_hex" | xxd -r -p)
+
+    local idx
+    for idx in "${!key_byte_lengths[@]}"; do
+        local key_byte_len="${key_byte_lengths[idx]}"
+        local value_byte_len="${value_byte_lengths[idx]}"
+        local decoded_key decoded_value
+
+        LC_ALL=C read -r -N "$key_byte_len" decoded_key <&3
+        LC_ALL=C read -r -N "$value_byte_len" decoded_value <&3
+
+        [[ "$decoded_value" == "nul" ]] && decoded_value=""
+        octopus_parameters["$decoded_key"]="$decoded_value"
+    done
+
+    exec 3<&-
+}
+
+bashParametersArrayFeatureToggle=#### BashParametersArrayFeatureToggle ####
+
+if [ "$bashParametersArrayFeatureToggle" = true ]; then
+    if (( ${BASH_VERSINFO[0]:-0} > 4 || (${BASH_VERSINFO[0]:-0} == 4 && ${BASH_VERSINFO[1]:-0} > 2) )); then
+        if command -v xxd > /dev/null; then
```

Comment:
We don't have to use xxd, but if we don't the implementation is very messy and I had to write the hex string to disk for performance issues (It was being cleaned up immediately after). Implementation on [this commit ](https://github.com/OctopusDeploy/Calamari/pull/1441/files/ad7439ddc15ca0b112b47cd2a8c69b737ec80c67)